src/olympia/blocklist/mlbf.py

```diff
@@ -72,178 +74,210 @@ def fetch_all_versions_from_db(excluding_version_ids=None):
     qs = Version.unfiltered.exclude(id__in=excluding_version_ids or ()).values_list(
         'addon__addonguid__guid', 'version'
     )
-    return list(qs)
+    return set(qs)
+
+
+class BaseMLBFLoader:
+    def __init__(self, storage: SafeStorage, _cache_path: str):
+        self.storage = storage
+        self._cache_path = _cache_path
+
+    @cached_property
+    def _raw(self):
+        """
+        raw serializable data for the given MLBFLoader.
+        """
+        return {key.value: self[key] for key in MLBFDataType}
+
+    def __getitem__(self, key: MLBFDataType) -> List[str]:
+        return getattr(self, f'{key.value}_items')
+
+    @cached_property
+    def cache_path(self):
+        return self._file_path(name=self.CACHE_FILE, extension='json')
+
+    @cached_property
+    def blocked_items(self):
+        raise NotImplementedError
+
+    @cached_property
+    def not_blocked_items(self):
+        raise NotImplementedError
+
+
+class MLBFStorageLoader(BaseMLBFLoader):
+    def __init__(self, storage: SafeStorage, _cache_path: str):
+        super().__init__(storage, _cache_path)
+        with self.storage.open(self._cache_path, 'r') as f:
+            self._data = json.load(f)
+
+    @cached_property
+    def blocked_items(self):
+        return self._data.get(MLBFDataType.BLOCKED.value)
+
+    @cached_property
+    def not_blocked_items(self):
+        return self._data.get(MLBFDataType.NOT_BLOCKED.value)
+
+
+class MLBFDataBaseLoader(BaseMLBFLoader):
+    def __init__(self, storage: SafeStorage, _cache_path: str):
+        super().__init__(storage, _cache_path)
+        self._version_excludes = []
+
+        # TODO: there is an edge case where you create a new filter from
+        # a previously used time stamp. THis could lead to invalid files
+        # a filter using the DB should either clear the storage files
+        # or raise to not allow reusing the same time stamp.
+        # it is possibly debatable whether you should be able to
+        # determine the created_at time as an argument at all
```

Comment:
It would lead to really corrupted data but yeah probably unlikely in practice. Just think it's odd to leave such a gaping whole to be covered onnly by the happenstance of time.